### ISSUE TITLE
Error when handling dynamic dimension other than the first

### ISSUE THREAD

**Issue by chentong319** [onnx/onnx-mlir#438]
There are 6 backend tests for concat Op: concatenate each dimension for 1D, 2D and 3D tensor. All passed except second and third dimension for 3D tensor (3D_axis_1 and 3D_axis_2). The MLIR output seems correct. I checked the LLVMIR output of 3D_axis_1 compared with that of 3D_axis_1.
The code for dimension is obviously wrong. Not clear which lowering after onnx to krnl went wrong.

Details:
The 3D_axis_1 MLIR code for output tensor is
```
   #map0 = affine_map<()[s0, s1] -> (s0 + s1)>

...
    %c1 = constant 1 : index
    %1 = dim %arg0, %c1 : memref<2x?x2xf32>
    %2 = dim %arg1, %c1 : memref<2x?x2xf32>
    %3 = affine.apply #map0()[%1, %2]
    %4 = alloc(%3) : memref<2x?x2xf32>
```

There are two arguments of which the description is put into %9 and %19. The input shape is <?x2x2xf32> for 3D_axis_0 and <2x?x2xf32> for 3D_axis_1.

For the correct 3D_axis_0, the first dimension size of input are loaded, added and store for the output. The other two dimension size are constant. Relative code is listed below:
```
    %23 = llvm.extractvalue %9[3, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %24 = llvm.extractvalue %19[3, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %25 = llvm.add %23, %24 : !llvm.i64
    %26 = llvm.mlir.constant(2 : index) : !llvm.i64
    %27 = llvm.mlir.constant(2 : index) : !llvm.i64
...
    %45 = llvm.insertvalue %25, %41[3, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %46 = llvm.insertvalue %44, %45[4, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %47 = llvm.insertvalue %26, %46[3, 1] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %48 = llvm.insertvalue %43, %47[4, 1] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %49 = llvm.insertvalue %27, %48[3, 2] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %50 = llvm.insertvalue %42, %49[4, 2] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
```

However, for 3D_axis_1, the code does not load the size of the second dimension of the two input. Instead, the first dimension size of first input is loaded. The code is like this:

```
    %23 = llvm.extractvalue %9[3, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %24 = llvm.mlir.constant(4 : index) : !llvm.i64
    %25 = llvm.mlir.constant(2 : index) : !llvm.i64
...
    %43 = llvm.insertvalue %23, %39[3, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %44 = llvm.insertvalue %42, %43[4, 0] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %45 = llvm.insertvalue %24, %44[3, 1] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %46 = llvm.insertvalue %41, %45[4, 1] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
    %47 = llvm.insertvalue %25, %46[3, 2] : !llvm.struct<(ptr<float>, ptr<float>, i64, array<3 x i64>, array<3 x i64>)>
```

**Comment by AlexandreEichenberger** [comment]
When I look at the code:

'''
 #map0 = affine_map<()[s0, s1] -> (s0 + s1)>

...
    %c1 = constant 1 : index
    %1 = dim %arg0, %c1 : memref<2x?x2xf32>
    %2 = dim %arg1, %c1 : memref<2x?x2xf32>
    %3 = affine.apply #map0()[%1, %2]
    %4 = alloc(%3) : memref<2x?x2xf32>
'''

is it possible that we should use the dim as opposed to the param dimensions for the map:

'''
 #map0 = affine_map<(d0, d1)[] -> (d0 + d1)>
'''

It seems more inline with the typical uses.

**Comment by AlexandreEichenberger** [comment]
@chentong319 would you be able to give it a quick try... that does not excuse the behavior you see, but if its a fix, then let's be it.

**Comment by tungld** [comment]
@chentong319 @AlexandreEichenberger Fixed this issue in PR #428, with this commit: https://github.com/onnx/onnx-mlir/pull/428/commits/914d870f5a274648d77e50eedaf10dd29ae2a100
